### PR TITLE
Fixing various functions in the file state module that use user.info

### LINKED ISSUE #57786: [BUG] file.symlink requires group attribute on proxy minion
**Description**
Creation of a symlink fails on a proxy minion when the optional parameter `group` is not specified.

**Setup**
I run a salt-master with several salt-proxy instances for Junos devices. As part of my state file, I wish to create a symlink locally on the salt-master.

**Steps to Reproduce the behavior**
```
# pillar/my-juniper.sls
proxy:
  proxytype: junos
  host: 10.0.0.1
  username: my_user
  port: 830
  password: my_password
```
```
# state.sls
create_symlink:
  file.symlink:
    - name: /testlink
    - target: /etc/hosts
```
`# salt my-juniper state.sls test`

This throws:
```
----------
          ID: create_symlink
    Function: file.symlink
        Name: /testlink
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python3.6/site-packages/salt/state.py", line 1981, in call
                  **cdata['kwargs'])
                File "/usr/lib/python3.6/site-packages/salt/loader.py", line 1977, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python3.6/site-packages/salt/states/file.py", line 1727, in symlink
                  __salt__['user.info'](user).get('gid', 0)
                File "/usr/lib/python3.6/site-packages/salt/loader.py", line 1248, in __getitem__
                  func = super(LazyLoader, self).__getitem__(item)
                File "/usr/lib/python3.6/site-packages/salt/utils/lazy.py", line 108, in __getitem__
                  raise KeyError(key)
              KeyError: 'user.info'
     Started: 15:07:39.056091
    Duration: 6769.352 ms
     Changes:   
```

When adding the optional `group` parameter, the state runs successfully:
```
create_symlink:
  file.symlink:
    - name: /testlink
    - target: /etc/hosts
    - group: root
```
The reason for this is that the code depends on `user.info`, which is not loaded:
```
# salt my-juniper user.info root
my-juniper:
    'user.info' is not available.
ERROR: Minions returned with non-zero exit code
```

**Expected behavior**
A symlink can be created without specifying the group.

**Screenshots**
See above.

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
Salt Version:
           Salt: 3000.3
 
Dependency Versions:
           cffi: 1.14.0
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.11.2
        libgit2: Not Installed
       M2Crypto: 0.35.2
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: 2.20
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.8 (default, Apr  2 2020, 13:34:55)
   python-gnupg: Not Installed
         PyYAML: 5.3.1
          PyZMQ: 15.3.0
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7.8.2003 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-1127.8.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.8.2003 Core
```
</details>

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Fixing various functions in the file state module that use user.info to get group information, certain hosts particularly proxy minions do not have the user.info function avaiable.  In those cases we fall back to using the username for the group name.  Also moving the existing file state module unit tests over to pytest and breaking them out into seperate files for easier management.

### What issues does this PR fix or reference?
Fixes: #57786

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: tests/pytests/unit/states/file/test_private_functions.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            root_tmp_dir,
+            dir_mode=oct(expected_mode),
+            file_mode=oct(expected_mode),
+            recurse=["mode"],
+        )
+        assert changed_files == set(ret[-1].keys())
+
+    finally:
+        # Cleanup
+        shutil.rmtree(root_tmp_dir)
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/file/test_private_functions.py:102]:
```suggestion
def test__check_directory(tmp_path):
    """
    Test the _check_directory function
    Make sure that recursive file permission checks return correctly
    """
    # set file permissions
    # Run _check_directory function
    # Verify that it returns correctly
    # Delete tmp directory structure
    root_tmp_dir = str(tmp_path / "test__check_dir")
    expected_mode = 0o770
    changed_mode = 0o755
    depth = 3
    def create_files(tmp_dir):
        for f in range(depth):
            path = os.path.join(tmp_dir, "file_{:03}.txt".format(f))
            with salt.utils.files.fopen(path, "w+"):
                os.chmod(path, expected_mode)

    # Create tmp directory structure
    os.mkdir(root_tmp_dir)
    os.chmod(root_tmp_dir, expected_mode)
    create_files(root_tmp_dir)

    for d in range(depth):
        dir_name = os.path.join(root_tmp_dir, "dir{:03}".format(d))
        os.mkdir(dir_name)
        os.chmod(dir_name, expected_mode)
        create_files(dir_name)
        for s in range(depth):
            sub_dir_name = os.path.join(dir_name, "dir{:03}".format(s))
            os.mkdir(sub_dir_name)
            os.chmod(sub_dir_name, expected_mode)
            create_files(sub_dir_name)
    # Symlinks on linux systems always have 0o777 permissions.
    # Ensure we are not treating them as modified files.
    target_dir = os.path.join(root_tmp_dir, "link_target_dir")
    target_file = os.path.join(target_dir, "link_target_file")
    link_dir = os.path.join(root_tmp_dir, "link_dir")
    link_to_dir = os.path.join(link_dir, "link_to_dir")
    link_to_file = os.path.join(link_dir, "link_to_file")

    os.mkdir(target_dir)
    os.mkdir(link_dir)
    with salt.utils.files.fopen(target_file, "w+"):
        pass
    os.symlink(target_dir, link_to_dir)
    os.symlink(target_file, link_to_file)
    for path in (target_dir, target_file, link_dir, link_to_dir, link_to_file):
        try:
            os.chmod(path, expected_mode, follow_symlinks=False)
        except (NotImplementedError, SystemError, OSError):
            os.chmod(path, expected_mode)

    # Set some bad permissions
    changed_files = {
        os.path.join(root_tmp_dir, "file_000.txt"),
        os.path.join(root_tmp_dir, "dir002", "file_000.txt"),
        os.path.join(root_tmp_dir, "dir000", "dir001", "file_002.txt"),
        os.path.join(root_tmp_dir, "dir001", "dir002"),
        os.path.join(root_tmp_dir, "dir002", "dir000"),
        os.path.join(root_tmp_dir, "dir001"),
    }
    for c in changed_files:
        os.chmod(c, changed_mode)

    ret = filestate._check_directory(
        root_tmp_dir,
        dir_mode=oct(expected_mode),
        file_mode=oct(expected_mode),
        recurse=["mode"],
    )
    assert changed_files == set(ret[-1].keys())
```
We could also use `pathlib.Path(...).{touch,mkdir,chmod}()` here too, but not a blocker.

File: tests/pytests/unit/states/file/test_rename.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    mock_lex = MagicMock(side_effect=[False, True, True])
+    with patch.object(os.path, "isabs", mock_f):
+        comt = "Specified file {} is not an absolute path".format(name)
+        ret.update({"comment": comt, "name": name})
+        assert filestate.rename(name, source) == ret
+
+    mock_lex = MagicMock(return_value=False)
+    with patch.object(os.path, "isabs", mock_t):
+        with patch.object(os.path, "lexists", mock_lex):
+            comt = 'Source file "{}" has already been moved out of ' "place".format(
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/file/test_rename.py:66]:
```suggestion
            comt = 'Source file "{}" has already been moved out of place'.format(
```

File: tests/pytests/unit/states/file/test_rename.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                source
+            )
+            ret.update({"comment": comt, "result": True})
+            assert filestate.rename(name, source) == ret
+
+    mock_lex = MagicMock(side_effect=[True, True, True])
+    with patch.object(os.path, "isabs", mock_t):
+        with patch.object(os.path, "lexists", mock_lex):
+            comt = 'The target file "{}" exists and will not be ' "overwritten".format(
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/file/test_rename.py:75]:
```suggestion
            comt = 'The target file "{}" exists and will not be overwritten'.format(
```

File: tests/pytests/unit/states/file/test_rename.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            )
+            ret.update({"comment": comt, "result": True})
+            assert filestate.rename(name, source) == ret
+
+    mock_lex = MagicMock(side_effect=[True, True, True])
+    mock_rem = MagicMock(side_effect=IOError)
+    with patch.object(os.path, "isabs", mock_t):
+        with patch.object(os.path, "lexists", mock_lex):
+            with patch.dict(filestate.__opts__, {"test": False}):
+                comt = 'Failed to delete "{}" in preparation for ' "forced move".format(
```
Comment (s0undt3ch) [inline tests/pytests/unit/states/file/test_rename.py:86]:
```suggestion
                comt = 'Failed to delete "{}" in preparation for forced move'.format(
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by Ch3LL — DISMISSED** [review]
yay for taking the time to migrate the file tests :) Looks good other than what pedro mentioned

**Review by Ch3LL — DISMISSED** [review]

**Review by Ch3LL — DISMISSED** [review]

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**